### PR TITLE
Support EntitySpec ConfigKeys type.

### PR DESCRIPTION
**Pull request by kiuby88**
Adding support for creating a EntitySpec from a ConfigKey

### CODE REVIEW

**Comment by kiuby88** [comment]
@sjcorbett  and @robertgmoss,
does it make sense for you?

Thx

File: brooklyn-tosca-transformer/src/test/resources/templates/dynamiccluster-with-memberspec.tosca.yaml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      type: org.apache.brooklyn.entity.webapp.ControlledDynamicWebAppCluster
+      properties:
+        dynamiccluster.memberspec:
+          $brooklyn:entitySpec:
+            type: org.apache.brooklyn.entity.webapp.jboss.JBoss7Server
+
+        http.port: "8080+"
+        wars.root: "http://search.maven.org/remotecontent?filepath=io/brooklyn/example/brooklyn-example-hello-world-sql-webapp/0.6.0/brooklyn-example-hello-world-sql-webapp-0.6.0.war"
+        java.sysprops:
+          brooklyn.example.db.url: $brooklyn:formatString("jdbc:%s%s?user=%s\\&password=%s", component("db").attributeWhenReady("datastore.url"), "visitors", "brooklyn", "br00k11n")
```
Comment (robertgmoss) [inline brooklyn-tosca-transformer/src/test/resources/templates/dynamiccluster-with-memberspec.tosca.yaml]: component db doesn't exist so blueprint fails to deploy

Comment (kiuby88) [inline brooklyn-tosca-transformer/src/test/resources/templates/dynamiccluster-with-memberspec.tosca.yaml]: Whoops.
Fixed.

**Comment by sjcorbett** [comment]
I think this PR is the wrong approach. It leaks Brooklyn into TOSCA:

``` yaml
    properties:
      dynamiccluster.memberspec:
        $brooklyn:entitySpec:
          type: org.apache.brooklyn.entity.webapp.jboss.JBoss7Server
```

Someone writing TOSCA shouldn't be thinking about `$brooklyn:entitySpec`.

I'd prefer to explore some different approaches.

One is to use the [scalable](http://docs.oasis-open.org/tosca/TOSCA-Simple-Profile-YAML/v1.0/csd04/TOSCA-Simple-Profile-YAML-v1.0-csd04.html#_Toc430015753) capability. I imagine that the transformer would recognise the capability and turn such specs into cluster+memberSpec.

Another idea is to represent both clusters and their members in the blueprint and to link them with a relationship. For example, something like:

``` yaml
    relationship_types:
      brooklyn.relationship.ClusteredOn:
        derived_from: tosca.relationships.HostedOn
```

Along with a node type using the relationship in a requirement with lower_bound and upper_bound both one. The transformer would recognise the relationship and configure the source as a member spec on the target. 

My preference is the second approach.

@kiuby88 What do you think?

@robertgmoss (and @ahgittin) do you have any additional comments?

**Comment by ahgittin** [comment]
Completely agree.  While it might be useful to permit this a tosca user would probably expect to define the entity spec as a tosca element which this node references somehow.

As an aside I'm not sure this is supported by the tosca spec.  Does it allow freeform maps?  I've only seen it allowing maps where the keys and values are completely described.

Returning to the thread, i think tosca assumes that _nodes_ are all deployed -- so you can't really include a spec as a node.  It has node_templates for that.  But node_templates have a couple problems I think:
- can we target a relationship at a node_template?
- how would we customize it, e.g. to make it aware of a database?

maybe we could define a new "NodeTemplateUsedAsASpecInANode" type consisting of:

```
type: <string>  # a node_template
properties: <map>  # permitted keys and values are dependent on "type"
```

and then we could have

```
properties:
  dynamiccluster.memberspec: <NodeTemplateUsedAsASpecInANode>
```

or maybe it could even be part of a relationship, `USES_TEMPLATE` and `dynamiccluster.memberspec` comes from that relationship ?

**Comment by kiuby88** [comment]
Thanks for your replies so interesting, @ahgittin and @sjcorbett.

Really, my proposal does not make sense with TOSCA philosophy. I am trying to model a cluster configkey and I used an property directly, I forgot the relationships.

Then, I agree with use a relationship to configure the property. I think you propose something like the following:

```
cluster:
      type: org.apache.brooklyn.entity.webapp.ControlledDynamicWebAppCluster
      properties:
        ...
nodeSpec:
  type: NodeTemplateUsedAsASpecInANode
      requirements:
        - childOf:
            node: cluster
            relationship: tosca.relationships.ConfigureSpecAsProperty
            properties:
              prop: dynamiccluster.memberspec
```

However, I think the requirement should be from cluster side (relation source). I think the requirement should be defined by the node which needs the entitySpec. This may allow to avoid new type definitions as `NodeTemplateUsedAsASpecInANode`. When the topology is processed each node is transformed in a EntitySpec. Then, when the relations are managed the target node is added to the source node as a property.

```
cluster:
      type: org.apache.brooklyn.entity.webapp.ControlledDynamicWebAppCluster
      properties:

      requirements:
        - memberSpec.req:
            node: nodeSpec
            relationship: tosca.relationships.ConfigureSpecAsProperty
            properties:
              prop: dynamiccluster.memberspec

nodeSpec:
  type: org.apache.brooklyn.entity.webapp.jboss.JBoss7Server
```

We do not need to define new types, we only need to add the requirements to the types which define  `ConfigKey<EntitySpec>`.

WDYT?

**Comment by ahgittin** [comment]
Like this direction.

But how is it known that cluster is to be deployed but nodeSpec is not (it
is just a template)?

**Comment by robertgmoss** [comment]
@kiuby88 do you want to do anything with this or can it be closed?

**Comment by kiuby88** [comment]
@robertgmoss I have closed this PR, I will be back here ASAP.
Thanks a lot for the replies

**Pull request closed without merging**